engine/src/test/java/org/camunda/bpm/engine/test/bpmn/tasklistener/TaskListenerEventLifecycleTest.java

```diff
@@ -0,0 +1,586 @@
+    assertThat(orderedEvents.size()).isEqualTo(2);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void shouldFireCompleteEventOnTaskCompletedInAssignmentListenerWhenTaskCreated() {
+    // given
+    BpmnModelInstance model = createModelWithTaskEventsRecorderOnAssignedUserTask(TRACKED_EVENTS,
+                                                                                  null,
+                                                                                  TaskListener.EVENTNAME_ASSIGNMENT,
+                                                                                  CompletingTaskListener.class);
+    testRule.deploy(model);
+    runtimeService.startProcessInstanceByKey("process");
+    Task task = taskService.createTaskQuery().singleResult();
+
+    // when
+    taskService.setAssignee(task.getId(), "gonzo");
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+    assertThat(orderedEvents.size()).isEqualTo(4);
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_UPDATE,
+                                              TaskListener.EVENTNAME_ASSIGNMENT,
+                                              TaskListener.EVENTNAME_COMPLETE);
+  }
+
+  @Test
+  public void shouldFireCreateEventBeforeTimeoutEventWhenTaskCreated() {
+    // given
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm");
+    Calendar now = Calendar.getInstance();
+    now.add(Calendar.HOUR, -1);
+    BpmnModelInstance model = Bpmn.createExecutableProcess("process")
+        .startEvent()
+          .userTask("task")
+            .camundaTaskListenerClass(TaskListener.EVENTNAME_CREATE, RecorderTaskListener.class)
+            .camundaTaskListenerClassTimeoutWithDate(TaskListener.EVENTNAME_TIMEOUT,
+                                                     RecorderTaskListener.class,
+                                                     sdf.format(now.getTime()))
+        .endEvent()
+        .done();
+    testRule.deploy(model);
+
+    // when
+    runtimeService.startProcessInstanceByKey("process");
+    testRule.waitForJobExecutorToProcessAllJobs(0L);
+
+    // then
+    List<String> orderedEvents = RecorderTaskListener.getOrderedEvents();
+    assertThat(orderedEvents.size()).isEqualTo(2);
+
+    // the TIMEOUT event will always fire after the CREATE event, since the Timer Job can't be
+    // picked up by the JobExecutor before it's committed. And it is committed in the same
+    // transaction as the task creation phase.
+    assertThat(orderedEvents).containsExactly(TaskListener.EVENTNAME_CREATE,
+                                              TaskListener.EVENTNAME_TIMEOUT);
+  }
+
+  @Test
+  public void testTimeoutTaskListenerCanceledWhenTaskCompleted() {
```

Comment:
for new test classes we should try to stay consistent with the test method naming and stick to "shouldXYZ"
```suggestion
  public void shouldCancelTimeoutTaskListenerWhenTaskCompleted
```